core/primitives-core/src/profile.rs

```diff
@@ -94,37 +65,170 @@ impl ProfileData {
+            cost => actions_array.get(borsh_action_index(cost)).copied().unwrap_or(0)
+        };
+        let wasm_ext_profile = enum_map! {
+            cost => ext_array.get(borsh_ext_index(cost)).copied().unwrap_or(0)
+        };
+
+        Ok(Self { actions_profile, wasm_ext_profile, wasm_gas })
     }
 }
 
-impl fmt::Debug for ProfileData {
+impl BorshSerialize for ProfileDataV2 {
+    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> Result<(), std::io::Error> {
+        let mut actions_costs: Vec<u64> = vec![0u64; ActionCosts::LENGTH];
+        for (cost, gas) in self.actions_profile.iter() {
+            actions_costs[borsh_action_index(cost)] = *gas;
+        }
+        BorshSerialize::serialize(&actions_costs, writer)?;
+
+        let mut ext_costs: Vec<u64> = vec![0u64; ExtCosts::LENGTH];
+        for (cost, gas) in self.wasm_ext_profile.iter() {
+            ext_costs[borsh_ext_index(cost)] = *gas;
+        }
+        BorshSerialize::serialize(&ext_costs, writer)?;
+
+        let wasm_cost: u64 = self.wasm_gas;
+        BorshSerialize::serialize(&wasm_cost, writer)
+    }
+}
+
+/// Fixed index of an action cost for borsh (de)serialization.
+///
+/// We use borsh to store profiles on the DB and borsh is quite fragile with
+/// changes. This mapping is to decouple the Rust enum from the borsh
+/// representation. The enum can be changed freely but here in the mapping we
+/// can only append more values at the end.
+///
+/// TODO: Consider changing this to a different format (e.g. protobuf) because
+/// canonical representation is not required here.
+const fn borsh_action_index(action: ActionCosts) -> usize {
+    match action {
+        ActionCosts::create_account => 0,
+        ActionCosts::delete_account => 1,
+        ActionCosts::deploy_contract_base => 2,
+        ActionCosts::deploy_contract_byte => 3,
+        ActionCosts::function_call_base => 4,
+        ActionCosts::function_call_byte => 5,
+        ActionCosts::transfer => 6,
+        ActionCosts::stake => 7,
+        ActionCosts::add_full_access_key => 8,
+        ActionCosts::add_function_call_key_base => 9,
+        ActionCosts::add_function_call_key_byte => 10,
+        ActionCosts::delete_key => 11,
+        ActionCosts::new_action_receipt => 12,
+        ActionCosts::new_data_receipt_base => 13,
+        ActionCosts::new_data_receipt_byte => 14,
+    }
+}
+
+/// Fixed index of an ext cost for borsh (de)serialization.
+///
+/// We use borsh to store profiles on the DB and borsh is quite fragile with
+/// changes. This mapping is to decouple the Rust enum from the borsh
+/// representation. The enum can be changed freely but here in the mapping we
+/// can only append more values at the end.
+///
+/// TODO: Consider changing this to a different format (e.g. protobuf) because
+/// canonical representation is not required here.
+const fn borsh_ext_index(ext: ExtCosts) -> usize {
+    match ext {
+        ExtCosts::base => 0,
```

Comment:
We could keep the enum discriminants explicit and stable by specifying them in the enum declaration too, as such:

```
enum ExtCosts {
    base = 0,
    constract_loading_base = 1,
    ...
}
```

It isn’t clear to me whether this mapping approach achieves much over that other approach. But this doesn’t seem all too onerous either, so maybe it is okay to leave this as-is.

Reply:
I liked the discriminants approach and changed it.